calc/rule_scanner.go

```diff
@@ -271,78 +256,80 @@ func ruleToParsedRule(rule *model.Rule) (parsedRule *ParsedRule, allTagOrSels []
 	return
 }
 
-func extractTagsAndSelectors(rule *model.Rule) (src, dst, notSrc, notDst []tagOrSel, err error) {
+func extractTagsAndSelectors(rule *model.Rule) (src, dst, notSrc, notDst []selector.Selector, err error) {
+	var sel selector.Selector
 	if rule.SrcTag != "" {
-		src = append(src, tagOrSelFromTag(rule.SrcTag))
+		sel, err = selFromTag(rule.SrcTag)
+		if err != nil {
+			return
+		}
+		src = append(src, sel)
 	}
 	if rule.DstTag != "" {
-		dst = append(dst, tagOrSelFromTag(rule.DstTag))
+		sel, err = selFromTag(rule.DstTag)
+		if err != nil {
+			return
+		}
+		dst = append(dst, sel)
 	}
 	if rule.NotSrcTag != "" {
-		notSrc = append(notSrc, tagOrSelFromTag(rule.NotSrcTag))
+		sel, err = selFromTag(rule.NotSrcTag)
+		if err != nil {
+			return
+		}
+		notSrc = append(notSrc, sel)
 	}
 	if rule.NotDstTag != "" {
-		notDst = append(notDst, tagOrSelFromTag(rule.NotDstTag))
+		sel, err = selFromTag(rule.NotDstTag)
+		if err != nil {
+			return
+		}
+		notDst = append(notDst, sel)
 	}
-	var tos tagOrSel
 	if rule.SrcSelector != "" {
-		tos, err = tagOrSelFromSel(rule.SrcSelector)
+		sel, err = selector.Parse(rule.SrcSelector)
 		if err != nil {
 			return
 		}
-		src = append(src, tos)
+		src = append(src, sel)
 	}
 	if rule.DstSelector != "" {
-		tos, err = tagOrSelFromSel(rule.DstSelector)
+		sel, err = selector.Parse(rule.DstSelector)
 		if err != nil {
 			return
 		}
-		dst = append(dst, tos)
+		dst = append(dst, sel)
 	}
 	if rule.NotSrcSelector != "" {
-		tos, err = tagOrSelFromSel(rule.NotSrcSelector)
+		sel, err = selector.Parse(rule.NotSrcSelector)
 		if err != nil {
 			return
 		}
-		notSrc = append(notSrc, tos)
+		notSrc = append(notSrc, sel)
 	}
 	if rule.NotDstSelector != "" {
-		tos, err = tagOrSelFromSel(rule.NotDstSelector)
+		sel, err = selector.Parse(rule.NotDstSelector)
 		if err != nil {
 			return
 		}
-		notDst = append(notDst, tos)
+		notDst = append(notDst, sel)
 	}
 	return
 }
 
-type tagOrSel struct {
-	tag      string
-	selector selector.Selector
-	uid      string
-}
-
-func tagOrSelFromTag(tag string) tagOrSel {
-	return tagOrSel{tag: tag, uid: hash.MakeUniqueID("t", tag)}
-}
-
-func tagOrSelFromSel(sel string) (tos tagOrSel, err error) {
-	parsedSel, err := selector.Parse(sel)
-	if err == nil {
-		tos = tagOrSel{selector: parsedSel, uid: parsedSel.UniqueId()}
-	}
-	return
+func selFromTag(tag string) (selector.Selector, error) {
+	return selector.Parse(fmt.Sprintf("has(%s)", tag))
```

Comment:
We panic further up the stack.  This was some of my early Go code.  I'll clean it up...